src/java.base/share/classes/jdk/internal/loader/URLClassPath.java

```diff
@@ -177,12 +177,13 @@ public URLClassPath(URL[] urls, @SuppressWarnings("removal") AccessControlContex
      * Constructs a URLClassPath from a class path string.
      *
      * @param cp the class path string
+     * @param jarHandler the {@link URLStreamHandler} for {@code jar} protocol
      * @param skipEmptyElements indicates if empty elements are ignored or
      *        treated as the current working directory
      *
      * @apiNote Used to create the application class path.
      */
-    URLClassPath(String cp, boolean skipEmptyElements) {
+    URLClassPath(String cp, URLStreamHandler jarHandler, boolean skipEmptyElements) {
```

Comment:
This is a special constructor for the application class path so shouldn't have the jarHandle parameter. This should allow you to drop the changes to ClassLoaders.